### ISSUE TITLE
ci.stage.build2.org CI down?

### ISSUE THREAD

**Issue by hazelnusse** [build2/build2#15]
I was following along the steps of the [CppCon 2018: “C++ Dependency Management: from Package Consumption to Project Development”](https://youtu.be/Nni2Qu2WitY). At the `bdep ci` step I get:

```
$ bdep ci
submitting:
  to:      https://ci.stage.build2.org
  in:      https://github.com/hazelnusse/genuuid.git#master@04553327ef6c7d73c28dbcf35befd9222126542f
  package: genuuid
  version: 0.1.0-a.0.20190223012538.04553327ef6c                                         
continue? [y/n] y
################################################################################### 100.0%
CI request is queued: https://ci.stage.build2.org/@ff1cb6a8-293f-4c9d-9753-00b4ccd8885d  
reference: ff1cb6a8-293f-4c9d-9753-00b4ccd8885d
```
When I visit the URL, I see no builds. I tried refreshing multiple times, but nothing shows up:
![no_builds](https://user-images.githubusercontent.com/63265/53280166-0cf32380-36cc-11e9-95b8-c824059cc3c0.png)

The CI submission URL presented to me after running `bdep ci` is different from the one in shown in the CppCon talk (ci.cppget.org) -- I'm not sure why but perhaps because I'm running a pre-release verison of build2? Here are the versions of the build2 tools I have installed:

```
$ b --version
build2 0.9.0-a.0.3f32c45ea6e4
libbutl 0.9.0-a.0.239c198350b2
host x86_64-linux-gnu
Copyright (c) 2014-2019 Code Synthesis Ltd
This is free software released under the MIT license.
$ bdep --version
bdep 0.9.0-a.0.1e3ff99415d6
libbpkg 0.9.0-a.0.f88510630354
libbutl 0.9.0-a.0.239c198350b2
Copyright (c) 2014-2019 Code Synthesis Ltd
This is free software released under the MIT license.
$ bpkg --version
bpkg 0.9.0-a.0.fad62fd77012
libbpkg 0.9.0-a.0.f88510630354
libbutl 0.9.0-a.0.239c198350b2
Copyright (c) 2014-2019 Code Synthesis Ltd
This is free software released under the MIT license.
```

Somewhat related: Are there any example projects on Github or Gitlab that use travis-ci or gitlab-ci with build2? I guess you'd have to build and install build2 as part of the CI, right?

**Comment by hazelnusse** [comment]
I just noticed that 0.9 was released. When I use the updated version, I'm getting a build immediately:
https://ci.cppget.org/@2852453a-6ceb-40e4-bda5-c08a4ee36e45

So maybe this is expected behavior?

**Comment by boris-kolpackov** [comment]
Yes, staged pre-releases by default send CI requests to `ci.stage.build2.org`, which is down for some hardware upgrades.

Regarding travis/gitlab CI services, I am not aware of any examples. The idea, of course, is that you would have a much smoother experience if you use the `build2` CI services (though, admittedly, they are probably not as general-purpose, at least at the moment). What is the use-case you are looking to handle?

**Comment by hazelnusse** [comment]
Hi Boris, thanks for the response and the info on the hardware upgrades. A couple of things I was wondering about:
- Triggering CI upon PR creation or similar (all I've seen is `bdep ci` but maybe alternatives are documented and I just haven't read up on that yet).
- Build status indicators visible through links in README.md -- is this currently possible with existing build2 CI services?
- What is in involved in adding a toolchain/platform? In particular I am interested in embedded, so things like like https://developer.arm.com/open-source/gnu-toolchain/gnu-rm/downloads or similar would be great to have as part of CI.

**Comment by boris-kolpackov** [comment]
Yes, `bdep-ci(1)` is just one way to submit CI requests. For example, you could just do it manually via an HTML form (https://ci.cppget.org/?ci). The CI service protocol is documented here:

https://build2.org/brep/doc/build2-repository-interface-manual.xhtml#ci

While the only CI result output that is currently supported is what you see on ci.cppget.org, we've been thinking about supporting a more structured/machine-readable output (we were thinking about it in the context of `bdep-ci` "polling mode" where you see the result in the terminal instead of a web browser; but obviously this could be useful in other contexts as well). We could also probably produce something simpler, like a status (`error`, `warning`, `success`, etc) or even return a corresponding image -- I am not sure what would be the easiest to integrate/embed into a `README.md`. So we are quite open to extending/improving things in this area.

Finally, regarding adding a new build machine (which are QEMU/KVM VMs), there are several possibilities. If you want to run your own CI service then you can obviously have anything you want. If you want to make it available as part of, say, `ci.cppget.org`, then someone will need to provide a build host. This can range from you providing/maintaining the VM and us running it on our build hosts (we could probably do this for some more commonly-useful toolchains) to you providing a build host yourself and we just register it with `ci.cppget.org`. Again, this is all leading edge/open to explore. You can read more about the CI infrastructure here:

https://build2.org/buildos/doc/build2-buildos-manual.xhtml

**Comment by hazelnusse** [comment]
Thanks Boris. Let me do some more reading on the CI stuff and get back to you so we can have a more productive discussion.

Regarding the integration / embedding of the build result into the README.md, the way this is done with travis-ci is documented here:
https://docs.travis-ci.com/user/status-images/

And here is an example of how it looks in Github (along with other status images):
https://github.com/brndnmtthws/conky/blob/master/README.md

Basically, it is a line in the .md file that looks like:
[![Alt text here](Build status image URL here)](link to build status page here)

gitlab-ci does it slightly differently, full discussion and documentation here:
https://gitlab.com/gitlab-org/gitlab-ce/issues/41174
https://docs.gitlab.com/ce/api/project_badges.html

**Comment by boris-kolpackov** [comment]
Thanks for the pointers. My immediate question after reading through the GitHub description is how to communicate the CI task id to the status service. My understanding is that with GitHub the best one can get is the latest build status of a branch (which is not necessarily the status of the latest commit).

I see GitLab has variable substitutions, including branch and commit-id -- while not exactly how we identify things, we can probably work with that (via some kind of repository+snapshot to task-id reverse lookup).

**Comment by Klaim** [comment]
> Somewhat related: Are there any example projects on Github or Gitlab that use travis-ci or gitlab-ci with build2? I guess you'd have to build and install build2 as part of the CI, right?

As of recently there is a Github Action for installing build2 on CI: https://github.com/build2/setup-build2-github-action
